### PR TITLE
Optimize parquet IO efficiency

### PR DESCRIPTION
**Pull request by liuneng1994**
This PR has three changes:
1. Support hdfs readbigat.
2. Parallelize parquet row group reading.
3. New prefetch algorithm, hides IO overhead and reduces overall latency when single-threaded parquet parsing

image
The main principle is to prefetch the next row group while processing the current row group, hiding the IO time of the next row group. In remote data source scenarios such as HDFS, the overall latency can be greatly reduced.

### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Supports parallel reading of parquet row groups and prefetching of row groups in single-threaded mode

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by liuneng1994** [comment]
@nickitat Hi. Could you please add the `can be test` tag?

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 541268d524a9c5aa5c48addcb79eb8b4e7d014ce with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/69862/541268d524a9c5aa5c48addcb79eb8b4e7d014ce/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/69857/019b227edfc56d225ee87d4664714268a2ad3289/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Comment by nickitat** [comment]
@liuneng1994 could you share any benchmarking results

**Comment by liuneng1994** [comment]
> @liuneng1994 could you share any benchmarking results

I am currently on vacation. I will provide some performance test data later. This optimization has a better effect under the high network latency of HDFS. There seem to be some problems with the pread interface of HDFS. Assert failed will occur in debug mode.

File: src/Processors/Formats/Impl/ParquetBlockInputFormat.cpp [hunk 1]
```diff
@@ -656,6 +688,29 @@ void ParquetBlockInputFormat::threadFunction(size_t row_group_batch_idx)
...
+{
+    return max_decoding_threads == 1 && max_io_threads > 0 && format_settings.parquet.enable_row_group_prefetch && !format_settings.parquet.use_native_reader;
+}
+
+std::shared_ptr<arrow::RecordBatchReader> ParquetBlockInputFormat::RowGroupPrefetchIterator::nextRowGroupReader()
+{
+    if (next_row_group_idx >= row_group_batch.row_groups_idxs.size()) return nullptr;
+    std::shared_ptr<arrow::RecordBatchReader> reader;
+    THROW_ARROW_NOT_OK(row_group_batch.file_reader->GetRecordBatchReader({row_group_batch.row_groups_idxs[next_row_group_idx]}, column_indices, &reader));
+    prefetchNextRowGroupReader();
```
Comment (nickitat) [inline src/Processors/Formats/Impl/ParquetBlockInputFormat.cpp:701]: could you clarify pls: do we prefetch group `g` before starting to read it or do we prefetch group `g+1` before starting to read group `g`?

**Comment by nickitat** [comment]
generally two things i'd think of: 
1. is it right to disable read coalescing? how we make sure we don't read too small batches if row groups are small
2. prefetch works if prefetched request is able to (almost) complete by the time data is needed. because of this we prefetch on ith iteration the data that is about to be processed on i+n-th iteration. so it makes sense to check if we prefetch with enough look ahead

**Comment by liuneng1994** [comment]
> generally two things i'd think of:
> 
> 1. is it right to disable read coalescing? how we make sure we don't read too small batches if row groups are small
> 2. prefetch works if prefetched request is able to (almost) complete by the time data is needed. because of this we prefetch on ith iteration the data that is about to be processed on i+n-th iteration. so it makes sense to check if we prefetch with enough look ahead

1. I missed the situation where the row group was too small and only merged the IO within the row group. I should reuse min_bytes_for_seek to merge small row groups.
2. For the first row group, a prefetch task will be initiated in the constructor. Row group g+1 will be prefetched after the RecordBatchReader of row group g is created. At this time, row group g has not yet started to be parsed. It is expected that row group g+1 will be downloaded while row group g is being parsed. I can add a new metric fetch_wait_time to observe whether prefetch is effective.

**Comment by liuneng1994** [comment]
a simple test:
dataset hits, fs minio, network 1000Mbps

enable prefetch, max_threads=1
image

disable prefetch, max_threads=1
image

In the above test, we can see that after enabling prefetch, the fetch wait time is 60s, the s3 download time is 130s, and the total query time is 130s. The parsing time of parquet is completely hidden, which means that downloading and parsing are parallel. The query time = max(download time, parsing time)

File: src/Processors/Formats/Impl/ParquetBlockInputFormat.cpp [hunk 2]
```diff
@@ -546,8 +559,19 @@ void ParquetBlockInputFormat::initializeRowGroupBatchReader(size_t row_group_bat
...
     // This adds one unnecessary copy. We should probably do coalescing and prefetch scheduling on
     // our side instead.
-    arrow_properties.set_pre_buffer(true);
-    auto cache_options = arrow::io::CacheOptions::LazyDefaults();
+    arrow::io::CacheOptions cache_options;
+
+    if (row_group_prefetch)
+    {
+        // Manual prefetch via RowGroupPrefetchIterator
+        arrow_properties.set_pre_buffer(false);
```
[the remark this answers is not in the document]
Comment (liuneng1994) [reply src/Processors/Formats/Impl/ParquetBlockInputFormat.cpp:574]: RowGroupPrefetchIterator will call file_reader->PreBuffer, and the internal implementation of PreBuffer will merge io

**Comment by liuneng1994** [comment]
need https://github.com/ClickHouse/libhdfs3/pull/68 for ci

**Comment by liuneng1994** [comment]
@nickitat any comments?

**Review by nickitat — APPROVED** [review]

**Comment by liuneng1994** [comment]
@nickitat Hi, Can you help merge this PR https://github.com/ClickHouse/libhdfs3/pull/68? I will update submodule hdfs after merging it. Otherwise, CI in debug mode will fail.

**Comment by nickitat** [comment]
> @nickitat Hi, Can you help merge this PR [ClickHouse/libhdfs3#68](https://github.com/ClickHouse/libhdfs3/pull/68)? I will update submodule hdfs after merging it. Otherwise, CI in debug mode will fail.

done

**Comment by liuneng1994** [comment]
input_format_parquet_filter_push_down can cause the bytes_reads results in the test cases in the bloom filter to occasionally deviate by a few bytes, so prefetch is disabled in the above use cases

**Comment by liuneng1994** [comment]
@nickitat ci passed, ready to merge

**Pull request merged**